src/ring_theory/ore_localization/ore_set.lean

```diff
@@ -0,0 +1,108 @@
+/-
+Copyright (c) 2022 Jakob von Raumer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jakob von Raumer, Kevin Klinge
+-/
+import group_theory.subgroup.basic
+
+/-!
+
+# (Right) Ore sets
+
+This defines right Ore sets on arbitrary monoids.
+
+## References
+
+* https://ncatlab.org/nlab/show/Ore+set
+
+-/
+
+namespace ore_localization
+
+section monoid
+
+/-- A submonoid `S` of a monoid `R` is (right) Ore if common factors on the left can be turned
+into common factors on the right, and if each pair of `r : R` and `s : S` admits an Ore numerator
+`v : R` and an Ore denominator `u : S` such that `r * u = s * v`. -/
+class ore_set {R : Type*} [monoid R] (S : submonoid R) :=
+(ore_left_cancel : ∀ (r₁ r₂ : R) (s : S), ↑s * r₁ = s * r₂ → {s' : S // r₁ * s' = r₂ * s'})
+(ore_num : R → S → R)
+(ore_denom : R → S → S)
+(ore_eq : ∀ (r : R) (s : S), r * ore_denom r s = s * ore_num r s)
```

Comment:
I had the three combined before, but found the current way more usable. If we need all three together we can use the lemma where it's packaged into sigma types. I also had `trunc` in there before, but I think in future it might be nice to actually reason about the independence of a choice of `num` and `denom`.

As for generalising `submonoid.localization_map`: We were hesitant to touch the current localization implementation of localization, because while Ore localization is used in non-commutative algebra, it's in no way seen as canon. But I'll take a look at `submonoid.localization_map`.

Reply:
What would it mean for R not to be commutative in an algebra? It still has to act commutatively, right?